service_capacity_modeling/interface.py

```diff
@@ -756,6 +762,122 @@ class CurrentClusters(ExcludeUnsetModel):
+    # Set only if the source of the buffer is not the direct component
+    source: Optional[str] = None
+
+
+class Buffers(ExcludeUnsetModel):
+    """Typical buffers (headroom) over the requirements to build into the system
+
+    Note that typically models make these choices, for example if a workload
+    requires 10 CPU cores, but the operator likes to build in 2x buffer for
+    variance (20 cores provisioned), they would express that as:
+        Buffer(
+            desired={
+                "compute": 2.0
+            }
+        )
+
+    There is a series of fallbacks here over common desired buffers:
+        compute -> default
+        cpu -> compute -> default
+        network -> compute -> default
+        disk -> storage -> default
+        memory -> storage -> default
+        memory-traffic -> memory -> default
+        memory-write-cache -> memory -> default
+        memory-read-cache -> memory -> default
+    """
+
+    # The default buffer if a specific buffer isn't known
+    # Models should prefer to document their precise buffers in desired
+    default: float = 1.5
+    # Desired compute, storage, cpu, memory, etc... buffers
+    desired: Dict[str, float] = {}
+    # Derive these buffers from current clusters or model context
+    derived: List[str] = []
+
+    def buffer_for_component(
+        self, component: str, current: Optional[CurrentClusterCapacity] = None
+    ) -> Buffer:
+        # flake8: noqa: C901
+        """Calculates buffer for a given component, handling fallbacks
+
+        Typical usage would be buffer("compute") to get the compute buffer
+        or buffer("storage") to get the storage buffer, but you can ask specific
+        resources like buffer("cpu").
+
+        Returns a tuple of the buffer expressed as a ratio (e.g. 2.0 for 2x buffer)
+        and the source of that buffer, e.g. "compute" if "cpu" was requested but
+        there is not explicit cpu buffer set by the model
+        """
+        buffers = dict(self.desired)
+        if isinstance(component, BufferComponent):
+            component = component.value
+
+        if current is not None:
+            # TODO (jolynch | 2025-02-15) Use current cluster when provided
+            #  If derived is set for the component, we should take buffers
+            #  from the current utilization not from desired. If a cluster is
+            #  happily running at 1.4x buffer the model shouldn't enforce its
+            #  default
+            for _ in self.derived:
+                pass
+
+        if component in buffers:
+            return Buffer(component=component, ratio=buffers[component])
+
+        # Fallbacks
+        def buffer_for(val: str) -> Buffer:
+            if isinstance(val, BufferComponent):
+                val = val.value
+            return Buffer(component=component, source=val, ratio=buffers[val])
+
+        fallbacks = {
+            "cpu": [BufferComponent.compute],
+            "network": [BufferComponent.compute],
```

Comment:
Intentional to not have a dedicated fallback for `BufferComponent.network`?

Reply:
Intentional yes, correct maybe.

I think network is generally going to scale with traffic (the `compute` buffer concept) - but you can always specify network specifically. I'm more concerned with memory which I have scaling first with storage but ... that doesn't seem right for e.g. stateless (where storage would just be logs, and compute would actually be better)